app/utils/callback/lava.py

```diff
@@ -381,225 +323,289 @@ def _store_lava_json(job_data, meta, base_path=utils.BASE_PATH):
-                raise e
-
+    job_data["lab_name"] = metadata.get("lab_name")
+    job_data["version"] = metadata.get("version")
+    job_data["boot_log_html"] = metadata.get("boot_log_html")
+    utils.make_path(dir_path)
     with open(file_path, "wb") as f:
         f.write(json.dumps(job_data))
 
 
-def _get_test_case(tests, names):
-    tests_by_name = {t['name']: t for t in tests}
-    for name in names:
-        login = tests_by_name.get(name)
-        if login:
-            return login
-
+def store_test_log(metadata, log):
+    """Parse and save test logs
 
-def _add_login_case(meta, cases, login_tc):
-    # ToDo: consolidate with _add_test_results
-    test_case = {
-        models.VERSION_KEY: "1.1",
-        models.TIME_KEY: "0.0",
-        models.NAME_KEY: "login",
-        models.STATUS_KEY: login_tc["result"],
-    }
-    test_case.update({k: meta[k] for k in TEST_CASE_GROUP_KEYS})
-    cases.append(test_case)
+    Parse the LAVA v2 log in YAML format and save it
+    as plain text and HTML.
+    """
 
+    dir_path = metadata[models.DIRECTORY_PATH]
+    suite = metadata[models.PLAN_KEY]
+    utils.LOG.info("Generating {} "
+                   "log files in {}".format(suite, dir_path))
+    file_name = "-".join([suite, metadata[models.DEVICE_TYPE_KEY]])
+    files = tuple(".".join([file_name, ext]) for ext in ["txt", "html"])
+    (metadata[models.BOOT_LOG_KEY],
+     metadata[models.BOOT_LOG_HTML_KEY]) = files
+    txt_path, html_path = (os.path.join(dir_path, f) for f in files)
+    utils.make_path(dir_path)
+    with codecs.open(txt_path, "w", "utf-8") as txt:
+        with codecs.open(html_path, "w", "utf-8") as html:
+            utils.lava_log_parser.run(log, metadata, txt, html)
+
+
+class LogFragmentsMixin(object):
+    LOGIN_CASE_END_PATTERN = re.compile(r'end:.*login-action.*')
+    TEST_CASE_SIGNAL_PATTERN = re.compile(
+        r'\<LAVA_SIGNAL_TESTCASE TEST_CASE_ID.+>')
+    SIGNAL_RECEIVED_PATTERN = re.compile(r'Received signal: '
+                                         r'<TESTCASE> TEST_CASE_ID=(\w+)')
+
+    def _add_log_fragments(self):
+        lines_map = []
+        for path, tc in self._test_case_iter():
+            tc = tc
+            log_end_line = tc.get('log_end_line')
+            if log_end_line:
+                new_end_line = self._adjust_log_end_line(int(log_end_line))
+                lines_map.append((tc, int(new_end_line)))
+            else:
+                utils.LOG.warn('Log end line number not found for {}'
+                               .format('.'.join(path)))
+        start_line = self.start_log_line
+        for tc, end_line in sorted(lines_map, key=lambda x: x[1]):
+            tc[models.LOG_LINES_KEY] = self.log[start_line: end_line]
+            start_line = end_line + 1 \
+                if end_line < len(self.log) else end_line
+
+    @property
+    def start_log_line(self):
+        line_number = self._get_log_line_number(self.LOGIN_CASE_END_PATTERN)
```

Comment:
Docstrings are especially useful for public methods and properties.

Reply:
Added docstring